homeassistant/components/recorder/__init__.py

```diff
@@ -0,0 +1,358 @@
+
+    There is also the run that covers point_in_time.
+    """
+    _verify_instance()
+
+    recorder_runs = get_model('RecorderRuns')
+    if point_in_time is None or point_in_time > _INSTANCE.recording_start:
+        return recorder_runs(
+            end=None,
+            start=_INSTANCE.recording_start,
+            closed_incorrect=False)
+
+    return query('RecorderRuns').filter(
+        (recorder_runs.start < point_in_time) &
+        (recorder_runs.end > point_in_time)).first()
+
+
+def setup(hass, config):
+    """Setup the recorder."""
+    # pylint: disable=global-statement
+    # pylint: disable=too-many-locals
+    global _INSTANCE
+    purge_days = config.get(DOMAIN, {}).get(CONF_PURGE_DAYS)
+
+    db_url = config.get(DOMAIN, {}).get(CONF_DB_URL, None)
+    if not db_url:
+        db_url = DEFAULT_URL.format(
+            hass_config_path=hass.config.path(DEFAULT_DB_FILE))
+
+    _INSTANCE = Recorder(hass, purge_days=purge_days, uri=db_url)
+
+    return True
+
+
+def query(model_name, *args):
+    """Helper to return a query handle."""
+    if isinstance(model_name, str):
+        return Session().query(get_model(model_name), *args)
+    return Session().query(model_name, *args)
+
+
+def get_model(model_name):
+    """Get a model class."""
+    from homeassistant.components.recorder import models
+
+    return getattr(models, model_name)
+
+
+def log_error(e, retry_wait=0, rollback=True,
+              message="Error during query: %s"):
+    """Log about SQLAlchemy errors in a sane manner."""
+    import sqlalchemy.exc
+    if not isinstance(e, sqlalchemy.exc.OperationalError):
+        _LOGGER.exception(e)
+    else:
+        _LOGGER.error(message, str(e))
+    if rollback:
+        Session().rollback()
+    if retry_wait:
+        _LOGGER.info("Retrying failed query in %s seconds", QUERY_RETRY_WAIT)
+        time.sleep(retry_wait)
+
+
+class Recorder(threading.Thread):
+    """A threaded recorder class."""
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, hass, purge_days, uri):
+        """Initialize the recorder."""
+        threading.Thread.__init__(self)
+
+        self.hass = hass
+        self.purge_days = purge_days
+        self.queue = queue.Queue()
+        self.quit_object = object()
+        self.recording_start = dt_util.utcnow()
+        self.db_url = uri
+        self.db_ready = threading.Event()
+        self.engine = None
+        self._run = None
+
+        def start_recording(event):
+            """Start recording."""
+            self.start()
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_recording)
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
+        hass.bus.listen(MATCH_ALL, self.event_listener)
+
+    def run(self):
+        """Start processing events to save."""
+        from homeassistant.components.recorder.models import Events, States
+        import sqlalchemy.exc
+
+        while True:
```

Comment:
I believe it should `while True` like in the original, as this is the event reading loop.